File: 2013-04-22/dtrace/custom/custom.c
#include <stdio.h>
#include "custom.h"
#include "custom_probes.h"

int matt_custom_action_begin(const char* string)
{
	if (MATTCUSTOM_ACTION_BEGIN_ENABLED()) {
		MATTCUSTOM_ACTION_BEGIN(string);
	}
	printf("matt_custom_action_begin('%s')\n", string);
	return 1;
}

int matt_custom_action_end()
{
	if (MATTCUSTOM_ACTION_END_ENABLED()) {
		MATTCUSTOM_ACTION_END();
	}
	printf("matt_custom_action_end\n");
	return 2;
}

File: _posts/2013-04-22/dtrace/custom/custom.rb
require 'ffi'


module MattCustom
  LIB_PATH = File.expand_path(File.join(File.dirname(__FILE__), 'custom.so'))
  puts "lib path: #{LIB_PATH}"
  
  extend FFI::Library
  
  begin
    # load library with FFI
    ffi_lib LIB_PATH
    
    # create methods
    attach_function :matt_custom_action_begin, [:string], :int
    attach_function :matt_custom_action_end, [], :int
  rescue LoadError
    puts "failed to load library"
    
    # create empty stub methods
    def matt_custom_action_begin(label); end
    def matt_custom_action_end; end
    module_function :matt_custom_action_begin
    module_function :matt_custom_action_end
  end
  
  
  # trace interface
  def trace(label, &block)
    MattCustom.matt_custom_action_begin(label)
    yield
  ensure
    MattCustom.matt_custom_action_end()
  end
  
  module_function :trace
end

puts "hello"
puts MattCustom.matt_custom_action_begin("there")
